File: 01-syntax/python/async.py
import asyncio
import time
from typing import Dict, Any, List

# Simulate async tasks that take different amounts of time
async def fetch_user_data(id: int) -> Dict[str, Any]:
    await asyncio.sleep(2)  # Simulate async operation
    return {
        "id": id,
        "name": f"User {id}",
        "email": f"user{id}@example.com"
    }

async def fetch_user_posts(id: int) -> Dict[str, Any]:
    await asyncio.sleep(1.5)  # Simulate async operation
    return {
        "userId": id,
        "posts": [f"Post 1 by user {id}", f"Post 2 by user {id}"]
    }

async def fetch_user_settings(id: int) -> Dict[str, Any]:
    await asyncio.sleep(1)  # Simulate async operation
    return {
        "userId": id,
        "theme": "dark",
        "notifications": True
    }

# Running tasks in parallel using asyncio.gather()
async def load_user_data_parallel(user_id: int) -> Dict[str, Any]:
    print("Starting parallel execution...")
    start_time = time.time()
    
    try:
        # Run all tasks in parallel
        user_data, user_posts, user_settings = await asyncio.gather(
            fetch_user_data(user_id),
            fetch_user_posts(user_id),
            fetch_user_settings(user_id)
        )
        
        end_time = time.time()
        print(f"All tasks completed in {end_time - start_time:.2f} seconds")
        
        return {
            "user": user_data,
            "posts": user_posts,
            "settings": user_settings
        }
    except Exception as e:
        print(f"Error loading user data: {e}")
        return {}

# Alternative: Using asyncio.create_task() for parallel execution
async def load_user_data_parallel_with_tasks(user_id: int) -> Dict[str, Any]:
    print("Starting parallel execution with create_task...")
    start_time = time.time()
    
    try:
        # Create tasks for parallel execution
        task1 = asyncio.create_task(fetch_user_data(user_id))
        task2 = asyncio.create_task(fetch_user_posts(user_id))
        task3 = asyncio.create_task(fetch_user_settings(user_id))
        
        # Wait for all tasks to complete
        user_data = await task1
        user_posts = await task2
        user_settings = await task3
        
        end_time = time.time()
        print(f"All tasks completed in {end_time - start_time:.2f} seconds")
        
        return {
            "user": user_data,
            "posts": user_posts,
            "settings": user_settings
        }
    except Exception as e:
        print(f"Error loading user data: {e}")
        return {}

# Sequential execution (slower)
async def load_user_data_sequential(user_id: int) -> Dict[str, Any]:
    print("Starting sequential execution...")
    start_time = time.time()
    
    user_data = await fetch_user_data(user_id)
    user_posts = await fetch_user_posts(user_id)
    user_settings = await fetch_user_settings(user_id)
    
    end_time = time.time()
    print(f"Sequential tasks completed in {end_time - start_time:.2f} seconds")
    
    return {
        "user": user_data,
        "posts": user_posts,
        "settings": user_settings
    }

# Main function to run examples
async def main():
    print("=== Python Async Parallel Example ===")
    
    # Run parallel example with gather
    result = await load_user_data_parallel(123)
    print(f"Parallel result: {result}")
    
    print("\n=== Python Async Parallel with Tasks Example ===")
    result2 = await load_user_data_parallel_with_tasks(456)
    print(f"Tasks result: {result2}")
    
    print("\n=== Python Async Sequential Example ===")
    result3 = await load_user_data_sequential(789)
    print(f"Sequential result: {result3}")

# Run the async examples
if __name__ == "__main__":
    asyncio.run(main()) 

File: 01-syntax/javascript/async.js
// Example of running async tasks in parallel

// Simulate async tasks that take different amounts of time
const fetchUserData = (id) => {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve({ id, name: `User ${id}`, email: `user${id}@example.com` });
        }, 2000);
    });
};

const fetchUserPosts = (id) => {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve({ userId: id, posts: [`Post 1 by user ${id}`, `Post 2 by user ${id}`] });
        }, 1500);
    });
};

const fetchUserSettings = (id) => {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve({ userId: id, theme: 'dark', notifications: true });
        }, 1000);
    });
};

// Running tasks in parallel using Promise.all()
async function loadUserDataParallel(userId) {
    console.log('Starting parallel execution...');
    const startTime = Date.now();
    
    try {
        const [userData, userPosts, userSettings] = await Promise.all([
            fetchUserData(userId),
            fetchUserPosts(userId),
            fetchUserSettings(userId)
        ]);
        
        const endTime = Date.now();
        console.log(`All tasks completed in ${endTime - startTime}ms`);
        
        return {
            user: userData,
            posts: userPosts,
            settings: userSettings
        };
    } catch (error) {
        console.error('Error loading user data:', error);
    }
}

// Example usage
loadUserDataParallel(123).then(result => {
    console.log('Final result:', result);
});

// Compare with sequential execution (slower)
async function loadUserDataSequential(userId) {
    console.log('Starting sequential execution...');
    const startTime = Date.now();
    
    const userData = await fetchUserData(userId);
    const userPosts = await fetchUserPosts(userId);
    const userSettings = await fetchUserSettings(userId);
    
    const endTime = Date.now();
    console.log(`Sequential tasks completed in ${endTime - startTime}ms`);
    
    return { user: userData, posts: userPosts, settings: userSettings };
}

// Uncomment to see sequential vs parallel difference
// loadUserDataSequential(456).then(result => {
//     console.log('Sequential result:', result);
// });
